src/type-layout.md

```diff
@@ -222,48 +223,165 @@ assert_eq!(std::mem::size_of::<SizeRoundedUp>(), 8);  // Size of 6 from b,
+struct MyEnumPayloadC { x: u32, y: u8 }
+```
+
+<span id="c-primitive-representation">Combining the `C` representation and a
+primitive representation is only defined for enums with fields. The primitive
+representation modifies the `C` representation by changing the representation of
+the tag, e.g. `MyEnumTag` in the previous example, to have the representation of
+the chosen primitive representation. So, if you chose the `u8` representation,
+then the tag would have a size and alignment of 1 byte. </span>
+
+> Note: This representation was designed for primarily interfacing with C code
+> that already exists matching a common way Rust's enums are implemented in
+> C. If you have control over both the Rust and C code, such as using C as FFI
+> glue between Rust and some third language, then you should use a
+> [primitive representation](#primitive-representation-of-enums-with-fields)
+> instead.
+
+### Primitive Representations
 
 The *primitive representations* are the representations with the same names as
 the primitive integer types. That is: `u8`, `u16`, `u32`, `u64`, `usize`, `i8`,
 `i16`, `i32`, `i64`, and `isize`.
 
-Primitive representations can only be applied to enumerations.
+Primitive representations can only be applied to enumerations, and have
+different behavior whether the enum has fields or no fields. It is an error
+for [zero-variant enumerations] to have a primitive representation.
+
+Combining two primitive representations together is unspecified.
 
-For [C-like enumerations], they set the size and alignment to be the same as the
-primitive type of the same name. For example, a C-like enumeration with a `u8`
-representation can only have discriminants between 0 and 255 inclusive.
+Combining the `C` representation and a primitive representation is described
+[above](#c-primitive-representation).
 
-It is an error for [zero-variant enumerations] to have a primitive
-representation.
+#### Primitive Fepresentation of Field-less Enums
 
-For all other enumerations, the layout is unspecified.
+For [field-less enums], they set the size and alignment to be the same as
+the primitive type of the same name. For example, a field-less enum with
+a `u8` representation can only have discriminants between 0 and 255 inclusive.
 
-Likewise, combining two primitive representations together is unspecified.
+#### Primitive Representation of Enums With Fields
+
+For enums with fields, the enum will have the same type layout a union with the
+`C` representation that's fields consist of structs with the `C` representation
+corresponding to each variant in the enum. The first field in each struct is
+the same field-less enum with the same primitive representation that is
+the enum with all fields in its variants removed and the rest of the fields
+consisting of the fields of the corresponding variant in the order defined in
+original enumeration.
+
+Because unions with non-copy fields aren't allowed, this representation can only
```

Comment:
This doesn't appear to be the case: https://play.rust-lang.org/?gist=6d0c155ebdffe17c6a3cd9683318209b&version=nightly